MAINT: Move set_module from numpy.core to numpy._utils

seberg opened the pull request:
Note sure if others agree, this creates a `numpy._utils` module, because `numpy.core` already imports the C stuff.
The reason was that I wanted to use `set_module` when moving exceptions (since they currently have that).  But `set_module` being defined in `overrides` needs C to be imported.  

So I thought, maybe we actually need a place to put things that don't need depend on anything else to avoid circular imports...

If nobody else likes the idea, I am happy to close and just use `object.__module__ = "numpy"`, or put it elsewhere (but I didn't have a good idea for that).

seberg committed:
MAINT: Move set_module to numpy.core to use without C import

seberg commented:
Note that I have not changed imports that are using `from ...overrides import something, set_module`, since they need the rest anyway.

mattip commented:
Maybe make this `numpy.lib._utils`?

seberg commented:
Yeah, I wanted to put it somewhere deeper, and I don't quite love this either (so happy for other ideas even duplicating the helper).
The problem is, I also want this to be available before any C import happens...  `lib` pulls in the public submodules, which pull in all of NumPy (circular imports).

mattip commented:
I guess `numpy._utils` makes sense. Perhaps we should move `numpy/compat/_inspect.py` there as well.

mattip commented:
Ahh, could we have a `numpy._utils` (maybe a different name?) directory, and in that have a `set_module.py` file? Then in the `__init__.py` it could say something about "A place for code that does not import from the rest of NumPy, to reduce circular dependencies"

seberg commented:
Yeah, a submodule/folder make sense, how about being super clear, like `_pydefs`?

`_globals` and `_inspect` could go there as well, then, yeah.  Maybe more?  I am not sure I love it, but a place to reduce circular imports seems like it likely makes sense anyway.

mattip commented:
> how about being super clear, like `_pydefs`

I am bad with names. We are looking for a place to put things with no dependencies. Do we want to definitely restrict ourselves to pure-python modules or can we also put some simple no-dependency c-extension module in the future? I think I prefer something in the direction of `utils` or `basic`

seberg committed:
MAINT: make `_utils` a full module

seberg committed:
MAINT: Move _inspect and _pep440 from compat to _utils

Note that unfortunately, compat does expose _inspect as well,
so the import remains (just the definition place moves).

seberg commented:
Hmmm, yeah, i don't the name `_pydefs` either anyway.  `_utils` seems fine to me, it is at least a place where you would look for these things?

Created a module, left `set_module` in the `__init__` for the moment.  Moved `_inspect` and `_pep440`.  `_inspect` is actuall imported to `np.compat`, but maybe it makes sense to move it anyway?

seberg committed:
STY: Fix indentation of _utils

mattip commented on numpy/compat/__init__.py:
```diff
@@ -8,9 +8,10 @@
 """
-from . import _inspect
+
+from .._utils import _inspect
```
Is this line required?

mattip reviewed (approved):
LGTM.

mattip commented:
I will put this in soon if there are no further comments

mattip commented:
Thanks @seberg
